### DISCUSSION TITLE
Faster loading times when image is opened for the first time

### DISCUSSION THREAD

**Discussion by informatter** [openseadragon/openseadragon#2271] [Q&A]
Hi guys,

I am trying to improve the loading time of an image the first time is opened. All my images are high resoultion, some of them might be taken with 50 MP cameras as well. I have seen that the docs suggest tiling an image as a recommended strategy for fetching and loading the images for better performance. Unfortunately at the moment, it's not viable for me to implement tiling, as it would require lots of changes in the back end for my app's existing functionality to still work

To open a single image I am simply using:

```
  this.openSeadragonViewer.open({
    type: "image",
    url: imageAsset.url, // AWS S3 pre-signed GET URL 
  });
```
The first time I access the page in my app which displays an image, it takes a few seconds to load. If I then exit the page and access it again, it loads much faster. So I am guessing the OSD is caching the data to make things more efficient on further runs. Obviously, if the browser is refreshed the caching is lost, and opening the image again will be slower.

Is there a way to improve the loading time when an image is opened for the first time, without the need to tile the image?
I am also curious to know, which part might be degrading performance, would it be fetching the image from S3, loading it into the viewer, or both?

Looking forward to hearing any suggestions and thank you for your time!

**Comment by iangilman** [comment]
Tiling definitely would help. Another option would be to have a low resolution version and a high resolution version and use LegacyTileSource (https://openseadragon.github.io/examples/tilesource-legacy/). That'll get you the low resolution image quicker, so you'll see something, but it'll still take a moment to get the high resolution. Note that the LegacyTileSource doesn't internally build a tile pyramid, unlike the ImageTileSource, which is what you're using currently... That might affect display performance, but I'm not sure. 

And yes, OSD does some caching, just in memory, and the browser caches on disk. 

I'm assuming that it's the loading from S3 that is the bottleneck. You can take a look in the browser inspector to see how long it's taking. I suppose it could also be related to building the pyramid internally... You can turn that off by adding `buildPyramid: false` to your TileSource just to see.

**Reply by informatter** [reply]
@iangilman 
Thank you for your response, I did 2 tests on incognito sessions regarding the image pyramid and checked the timing in the in Chrome dev tools, both took roughly the same amount of time.

- Setting `buildPyramid` to `false`:
```
  this.openSeadragonViewer.open({
    type: "image",
    url: imageAsset.url,
    buildPyramid: false
  });
```
yielded:
<img width="577" alt="no_image_pyramid" src="https://user-images.githubusercontent.com/47017130/212291528-0581b7eb-fbf9-4c91-bafb-89e0c5d3963f.png">

And

```
  this.openSeadragonViewer.open({
    type: "image",
    url: imageAsset.url,
  });
```

yielded:
<img width="579" alt="with_image_pyramid" src="https://user-images.githubusercontent.com/47017130/212291968-2f172108-52f7-48f9-a65e-776a41448e12.png">


Basically, both methods take roughly 1.3 seconds to for requesting and read the response ( downloading the content), which is not that bad. Does it seem like the remaining 1.82 seconds are spent by OSD loading the image into the viewer ?

Thanks

**Comment by iangilman** [comment]
It looks to me like those tests are showing network timing? The pyramid building wouldn't affect that; it would be something that happens after the download is complete. I'm actually not sure the best way to check it, beyond building your own copy of OSD with some `console.time` calls in the right places. I suppose another thing would be to use `console.time` to test how long it takes from starting the image load to when it first draws on the screen or something. That'll give you the overall timing that you're looking at, and of course, if it's different with the pyramid building, you would know. 

When you say "remaining 1.82 seconds" are you talking about the "queuing" portion of the chart you shared? That happens before the load, so it wouldn't be OSD loading the image into the viewer. I'm not sure what that part of the chart is about, but it certainly seems worth investigating! Maybe the next step is to find out what resource scheduling and queuing mean here (in this tool you're using, not related to OSD).
